### ISSUE TITLE
StackOverflowError generating code from Postgresql XML schema with ARRAY columns

### ISSUE THREAD

**Issue by geoffhardy** [jOOQ/jOOQ#9067]
### Expected behavior and actual behavior:
I'm trying to generate code from an information_schema.xml I generated using XMLGenerator from a Postgresql database. The schema contains a several INTEGER ARRAY columns. When it processes these columns a StackOverflowError is thrown. If I use the JavaGenerator directly from a database connection, it works fine.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):
My employer won't allow me to submit code to an external site, so MCVE is not possible.

Create Postgresql database with INTEGER ARRAY columns, e.g.,

    CREATE TABLE IF NOT EXISTS answer_def (
        answer_def_id   serial PRIMARY KEY,
        content         varchar(256) NOT NULL,
    );
    CREATE TABLE IF NOT EXISTS question_def (
        question_def_id serial PRIMARY KEY,
        content         varchar(256) NOT NULL,
        answer_def_ids  integer[]
    );

Create an XMLGenerator config file with Postgres connection parameters, e.g.

    <configuration xmlns="http://www.jooq.org/xsd/jooq-codegen-3.11.0.xsd">
    <jdbc>
        <driver>org.postgresql.Driver</driver>
        <url>jdbc:postgresql://HOST:5432/DBNAME</url>
        <user>USER</user>
    </jdbc>

    <generator>
        <name>org.jooq.codegen.XMLGenerator</name>
        <database>
            <name>org.jooq.meta.postgres.PostgresDatabase</name>
            <inputSchema>public</inputSchema>
            <includes>.*</includes>
        </database>
        <target>
            <packageName>com.twosigma.re.scorecard.db.jooq</packageName>
            <directory>java/src</directory>
        </target>
        <generate>
            <pojos>true</pojos>
            <pojosEqualsAndHashCode>true</pojosEqualsAndHashCode>
        </generate>
    </generator>
    </configuration>

Create a JavaGenerator config file that reads the XML schema output.

    <configuration xmlns="http://www.jooq.org/xsd/jooq-codegen-3.11.0.xsd">
    <generator>
        <name>org.jooq.codegen.JavaGenerator</name>
        <database>
            <name>org.jooq.meta.xml.XMLDatabase</name>
            <includes>.*</includes>
            <excludes>pg_.*|pgstat.*|postgres.*|tds_.*|pgrowlocks</excludes>
            <properties>
                <property>
                    <key>dialect</key>
                    <value>POSTGRES</value>
                </property>
                <property>
                    <key>xml-file</key>
                    <value>path/information_schema.xml</value>
                </property>
            </properties>
        </database>
        <target>
            <packageName>com.project.db.jooq</packageName>
            <directory>java/gensrc/jooq</directory>
        </target>
        <generate>
            <pojos>true</pojos>
            <pojosEqualsAndHashCode>true</pojosEqualsAndHashCode>
        </generate>
    </generator>
    </configuration>

Run code gen tool with XMLGenerator config file. Output is another information_schema.xml file. An example of an ARRAY column looks like:

    CLASSPATH=...
    java org.jooq.codegen.GenerationTool config/jooq_xmlgen0.xml

    <column>
            <table_catalog></table_catalog>
            <table_schema>public</table_schema>
            <table_name>question_def</table_name>
            <column_name>answer_def_ids</column_name>
            <data_type>ARRAY</data_type>
            <character_maximum_length>0</character_maximum_length>
            <numeric_precision>0</numeric_precision>
            <numeric_scale>0</numeric_scale>
            <ordinal_position>8</ordinal_position>
            <is_nullable>true</is_nullable>
    </column>

Run code gen tool with JavaGenerator config file. 

    CLASSPATH=...
    java org.jooq.codegen.GenerationTool config/jooq_xmlgen0.xml`

Crashes with StackOverflowError below.

    DEBUG AbstractTypedElementDefinition - Type mapping             : public.question_def.answer_def_ids with type ARRAY
    Exception in thread "main" java.lang.StackOverflowError
        at java.base/java.lang.String.length(String.java:658)
        at java.base/java.lang.String.startsWith(String.java:1428)
        at java.base/java.lang.String.startsWith(String.java:1470)
        at org.jooq.codegen.GenerationUtil.getArrayBaseType(GenerationUtil.java:373)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6037)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6042)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6042)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6042)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6042)
        at org.jooq.codegen.JavaGenerator.getType(JavaGenerator.java:6042)

### Versions:

- jOOQ: Pro 3.11.2
- Java: 11.0.4 (OpenJDK)
- Database (include vendor): Postgresql 10.3
- OS: Linux (Debian) 4.14.67
- JDBC Driver (include name if inofficial driver): posgresql jdbc 42.1.4

**Comment by lukaseder** [comment]
Thank you very much for your report. It appears that this is the same issue as #6646. I'll close this one as a duplicate.

Thanks for the additional details to help reproduce this issue!

**Comment by lukaseder** [comment]
I could not reproduce this issue with neither 3.12.0 nor 3.11.12 anymore. This is the `pom.xml` execution configuration I've used:

```xml
<execution>
    <id>generate-informationschemaarray</id>
    <phase>generate-sources</phase>
    <goals>
        <goal>generate</goal>
    </goals>
    <configuration>
        <generator>
            <database>
                <name>org.jooq.meta.xml.XMLDatabase</name>
                <properties>
                    <property>
                        <key>dialect</key>
                        <value>POSTGRES</value>
                    </property>
                    <property>
                        <key>xmlFile</key>
                        <value>${basedir}/src/main/resources/informationschemaarray.xml</value>
                    </property>
                </properties>
            </database>
            <target>
                <packageName>org.jooq.codegen.test.informationschemaarray.db</packageName>
            </target>
        </generator>
    </configuration>
</execution>
```

And this is a minimal schema definition:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<information_schema>
    <schemata>
        <schema>
            <schema_name>public</schema_name>
        </schema>
    </schemata>
    <tables>
        <table>
            <table_schema>public</table_schema>
            <table_name>t</table_name>
        </table>
    </tables>
    <columns>
        <column>
            <table_catalog></table_catalog>
            <table_schema>public</table_schema>
            <table_name>t</table_name>
            <column_name>a</column_name>
            
            <!-- [#9067] This type has been seen to produce StackOverflowErrors -->
            <data_type>array</data_type>
            <numeric_precision>0</numeric_precision>
            <numeric_scale>0</numeric_scale>
            <ordinal_position>1</ordinal_position>
            <is_nullable>true</is_nullable>
        </column>
    </columns>
</information_schema>
```

Would upgrading to 3.11.12 be an option for you? Can you confirm this issue still exists in 3.11.12?

Notice, an MCVE does not imply you have to upload actual code from your application. If there was a way to produce a minimal example, similar to the snippets you've shown here, but in a complete way, that would certainly help reproduce the issue.

**Comment by knutwannheden** [comment]
I was able to reproduce this with the current 3.12.0 master branch. Changing the following line:

https://github.com/jOOQ/jOOQ/blob/66a45669c4412b3629728a137ee23d76b345add5/jOOQ-codegen/src/main/java/org/jooq/codegen/XMLGenerator.java#L167

to read:

```java
         column.setDataType(type.getUserType());
```

Appears to fix this particular problem. But we will of course have to make sure that this is also correct in all other cases.

**Comment by lukaseder** [comment]
Great news! What was I missing in my test case?

**Comment by knutwannheden** [comment]
Here is my MCVE: https://github.com/knutwannheden/jOOQ-mcve-9067

**Comment by knutwannheden** [comment]
I hope the different style of MCVE doesn't cause you any trouble.

**Comment by lukaseder** [comment]
Finally looking into this now. I can reproduce it using the MCVE

**Comment by lukaseder** [comment]
Can't be reproduced with HSQLDB, where we generate:

```
<data_type>INTEGER ARRAY</data_type>
```

Instead of just `ARRAY`

**Comment by lukaseder** [comment]
The fix described here fixed the underlying issue in `XMLGenerator` (i.e. this one: https://github.com/jOOQ/jOOQ/issues/6653):  https://github.com/jOOQ/jOOQ/issues/9067#issuecomment-523815608

But it doesn't fix the `StackOverflowError`, which can still happen if someone exports a type `ARRAY` by any means and imports it again. This issue will only fix the `StackOverflowError`

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.5 (#11039)

**Comment by lukaseder** [comment]
The fix leads to significant code generation regressions. Let's review this more carefully...
